standards/gitcoinGrants948.md

```diff
@@ -0,0 +1,127 @@
+# [WIP] ERC948 Subscription Standard (v1)
+
+## Summary
+The following is a proposed standard for subscription payments with read and write specs.
+
+## Abstract
+This standard allows a user to set a recurring payment to a service provider. This recurring transaction is ideally executed by a third party agent, but at a minimum the service provider themselves can execute the payment. Either side of the subscription should be able cancel the agreement with a reasonable amount of forewarning to the other party. This standard or similar should be integrated with future smart wallets to enable subscription payments in the Ethereum ecosystem.
+
+## Motivation
+Recurring payments are a widely use monetization method in traditional software. Currently, no such pattern exists with desirable UX traits in the decentralized space. Creating such a pattern will unlock many new avenues for revenue and provide an alternative to the dominant rent seeking and token-valuation based monetization methods in the current ecosystem.
+
+### ERC948 Read
+
+```
+interface ERC165 {
+
+@notice Query if a contract implements an interface
+@param interfaceID The interface identifier, as specified in ERC-165
+@dev Interface identification is specified in ERC-165. This function uses less than 30,000 gas.
+@return `true` if the contract implements `interfaceID` and `interfaceID` is not 0xffffffff, `false` otherwise
+
+function supportsInterface(bytes4 interfaceID) external view returns (bool);
+
+}
+
+interface ERC948Read is ERC165 {
+
+@dev Return the details of a subscription
+@param subscriptionId - A unique representation of the subscription based on the details of the agreement
+
+function getSubscription(
+   bytes subscriptionId
+ )
+ public
+ view
+ returns (
+   address _destination,
+   address _recipient,
+   address _agent,
+   uint _agentRewardPct,
+   uint _valuePerPeriod,
+   uint _secondsPerTimePeriod,
```

Comment:
This is interesting choice, but it's impossible to set up a subscription which would be charged on the 1st of every month.

Reply:
Yes, it is impossible to have it fire on the first of each month. It is not clear to me how your implementation accomplishes either however. I would like to learn. It appears to me that your implementation also has a rigid amount of time during each time period which is allotted on the creation of the subscription. The alternative being to supply a timestamp from an outside source which seems vulnerable. Or perhaps there is an ethereum calendar library that I am not aware of? Would you mind explaining your solution to me, or perhaps point me to resources that explain your approach? I'm happy to get on a phone call as well. Thank you.

Reply:
The basic idea is to provide billing period using a specific time unit. Time unit could be an enumerated constant (eg. month(0), day(1), hour(2),...). 
So for instance, billing every week would look like: (... _timeUnit=1, _period=7 ...)

Reply:
This is the same as in our proposal. I think we should distinguish the actual **payment** and **funds availability** for the subscription. 
Why not leave this up to the provider on when he wants to execute the payment, like every day, every month etc. 
Using `valuePerPeriod` enables to define the cost of subscription and required amount of funds to be available for it without a need of implementing calendar functions (not trivial in Solidity)

Reply:
Provider can execute the payment at his own leisure that is true. However, subscriber still needs a strong guarantee that the payment won't occur before specified dates. That would require on-chain logic.

I don't think we need a complete calendar functionality with features like time zones and daylight savings... A days-per-month mapping with logic to handle leap years and some time unit transform functions would probably suffice?